Stabilize the "IP" feature

caass opened the pull request:
Latest Status Update: https://github.com/rust-lang/rust/pull/76098#issuecomment-760841554

-----

Continuation of #66584, since it seems to be dead. This is my first time contributing and I did my best to read through the rustc-dev-guide, but I'll be honest I kind of skipped around a little bit. If there's anything I need to do that hasn't been done yet in terms of process, please let me know and I'll do it 😄 

The old PR seemed to me (could be wrong) to be almost done, but there were two issues that were left unresolved. I [left a comment](https://github.com/rust-lang/rust/pull/66584#issuecomment-683348800) on the PR, but I'll copy/paste it here for convenience:

If I understand correctly, the blockers currently are:

1. There should be some checking for IPv4 addresses mapped it IPv6 space ([comment](https://github.com/rust-lang/rust/pull/66584#issuecomment-597543156))

[From the RFC](https://tools.ietf.org/html/rfc4291#section-2.5.5):

>    Two types of IPv6 addresses are defined that carry an IPv4 address in the low-order 32 bits of the address.  These are the "IPv4-Compatible IPv6 address" and the "IPv4-mapped IPv6 address".
>
> [...]
>
> The "IPv4-Compatible IPv6 address" was defined to assist in the IPv6 transition.  The format of the "IPv4-Compatible IPv6 address" is as follows:
> 
> | 80 bits | 16 bits | 32 bits |
> | - | - | - |
> | 0000..............................0000 | 0000 | IPv4 address |
> 
> [...]
>
>    A second type of IPv6 address that holds an embedded IPv4 address is defined. This address type is used to represent the addresses of IPv4 nodes as IPv6 addresses. The format of the "IPv4-mapped IPv6 address" is as follows:
> | 80 bits | 16 bits | 32 bits |
> | - | - | - |
> | 0000..............................0000 | FFFF | IPv4 address |

If we already have capability to check IPv4 addresses, I feel like it's almost trivial to implement these rules for IPv6 as well if we just need to pad it with some extra bytes? Famous last words, I know...

2. There's some concern over the need for both `is_unicast_link_local_strict` and `is_unicast_link_local` ([comment](https://github.com/rust-lang/rust/pull/66584#issuecomment-574036125))

From [the commit that introduced the change](https://github.com/rust-lang/rust/commit/1f0aa4043b6051602f016b61a393ef5981b5b831):

> RFC 4291 is a little unclear about what is a unicast link local address.
> According to section 2.4, the entire fe80::/10 range is reserved for
> these addresses, but section 2.5.3 defines a stricter format for such
> addresses.
>
> After a [discussion](https://github.com/rust-lang/rust/issues/27709#issuecomment-400370706) is has been decided to add a different method for
> each definition, so this commit:
> 
>   - renames is_unicast_link_local() into is_unicast_link_local_strict()
>   - relaxed the check in is_unicast_link_local()

From [section 2.4](https://tools.ietf.org/html/rfc4291#section-2.4):

> | Address type | Binary prefix | IPv6 notation | Section |
> |-|-|-|-|
> | Unspecified | 00...0 (128 bits) | ::/128 | 2.5.2 |
> | Loopback | 00...1 (128 bits) | ::1/128 | 2.5.3 |
> | Multicast | 11111111 | FF00::/8 | 2.7 |
> | **Link-Local unicast** | **1111111010** | **FE80::/10** | **2.5.6** |
> | Global Unicast | (everything else) |  |  |

Section [section 2.5.3](https://tools.ietf.org/html/rfc4291#section-2.5.3) seems...unrelated? It has to do with the loopback address. [Section 2.5.6](https://tools.ietf.org/html/rfc4291#section-2.5.6) however, is where the problem lies:

> Link-Local addresses are for use on a single link. Link-Local addresses have the following format:
> | 10 bits | 16 bits | 64 bits |
> | - | - | - |
> | 1111111010 | 0 | interface ID |

So the difference in the spec is this:

| Spec | CIDR | Start | End | Available |
|-|-|-|-|-|
| 2.4 | FE80::/10 | fe80:0:0:0:0:0:0:0 | febf:ffff:ffff:ffff:ffff:ffff:ffff:ffff | 2^118 |
| 2.5.6 | FE80::/64 | fe80:0:0:0:0:0:0:0 | fe80:0:0:0:ffff:ffff:ffff:ffff | 2^64 |

This does seem like an issue, so I did some more digging to figure out what's going on. I found [this draft](https://tools.ietf.org/html/draft-petrescu-6man-ll-prefix-len-05), which appears to have [died](https://datatracker.ietf.org/doc/draft-petrescu-6man-ll-prefix-len/). It does however point us in the right direction:

>    The RFC "IPv6 Address Archi" illustrates the format of the link-local addresses.  From the illustration it MAY be understood that the length of the link-local prefix is 10 bits of value 1111111010 and 54 0 bits.
> 
>    [IANA](https://www.iana.org/assignments/ipv6-address-space/ipv6-address-space.xhtml) lists the "IPv6 prefix", and "Address Block", to be "fe80::/10" on its website.  It is possible that in the future the IETF could decide to use the bits 11-53.
> 
>    The [RFC 2464](https://tools.ietf.org/html/rfc2464#section-5) "IPv6-over-Ethernet" states that the prefix for link-local addresses is "fe80::/64".
> 
>    [RFC 6874](https://tools.ietf.org/html/rfc6874#section-4), "Representing IPv6 Zone Identifiers in Address Literals and Uniform Resource Identifiers" specifies the link-local addresses to be under prefix "fe80::/10".

~I'm not an expert in reading IETF material -- far from it -- but I am vaguely aware of the passage of time. Most documents  linked here that specify /64 are more recent than those specifying /10, unless I've missed some -- which is totally possible. Additionally, there are countless secondhand sources (blogs, stackoverflow, etc) online all talking about how /64 is what's commonly used.~

~Again, not an expert. I posted all this info here so someone more versed than me can evaluate and decide what the best course of action is. Without further input, I'd personally go ahead with just the one `is_unicast_link_local` that validates against /64, since it's a more strict subset of /10, and if someone runs into trouble with running custom link-local addresses outside of /64 space I kind of feel like that is a problem outside the scope of the Rust programming language, and is something they should talk to their network administrator about.~

***Edit: Per further discussion [here](https://github.com/rust-lang/rust/pull/76098#issuecomment-705928605) and [here](https://github.com/rust-lang/rust/pull/76098#issuecomment-706916840), I've decided to simply remove the `_strict` check to bring Rust's functionality more in line with other similar features in other languages (e.g. Go, C, Python) as well as the Linux kernel.***

At the moment, this PR is just fast-forwarding the existing work done by @little-dude to be up-to-date with rust-lang/rust. I wanted to get some confirmation that this hasn't been implemented yet in some other way / i'm not duplicating work / this feature is still needed before I get started on the code. Thanks in advance for your time! :heart:

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @cramertj (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

caass commented:
r? @KodrAus 

hi! tagging you because you were pretty active on the old pr 😅

pickfire commented on library/std/src/net/ip.rs:
```diff
@@ -118,16 +110,81 @@ pub struct Ipv6Addr {
+/// [section 2 of IETF RFC 7346]: https://tools.ietf.org/html/rfc7346#section-2
 #[derive(Copy, PartialEq, Eq, Clone, Hash, Debug)]
+#[non_exhaustive]
+#[stable(feature = "ip", since = "1.42.0")]
```
`1.42.0`? Shouldn't this be `1.47.0`?
```suggestion
#[stable(feature = "ip", since = "1.47.0")]
```

caass replied:
Yes! Like I said my comment, I haven't actually written any code yet, I just wanted to continue the conversation and see if this was still something that was needed / wanted and make sure that I was clear to start working on it :)

pickfire commented on library/std/src/net/ip.rs:
```diff
@@ -1436,18 +1700,28 @@ impl Ipv6Addr {
+    /// [changes]: ../net/index.html#stability-guarantees-for-ietf-defined-behavior
+    #[stable(feature = "ip", since = "1.42.0")]
     pub fn multicast_scope(&self) -> Option<Ipv6MulticastScope> {
+        use Ipv6MulticastScope::*;
```
I think we should keep the `Ipv6MulticastScope` to be specific, I don't know if the `*` import would be good here.

KodrAus commented:
Thanks for taking the time to dig through the state of things nd put this together @caass!

I’ll spend some time digging through this and try follow along the rationale.

Also cc @nanpuyue who might be interested in this.

KodrAus commented:
Since the FCP on the previous PR didn’t complete I think we’ll need to start a fresh one once this is ready to go.

little-dude commented:
> Since the FCP on the previous PR didn’t complete I think we’ll need to start a fresh one once this is ready to go.

Yeah I apologize for the time reviewers lost on this =/ And thank you @caass for picking this up.

caass commented:
> Yeah I apologize for the time reviewers lost on this =/

@little-dude without your work I wouldn't have even known where to start -- not all is lost!

bors commented:
:umbrella: The latest upstream changes (presumably #75979) made this pull request unmergeable. Please resolve the merge conflicts.

KodrAus commented:
@little-dude Oh that's not at all a comment on you! You've done awesome work bringing this together.

KodrAus commented on library/std/src/net/ip.rs:
```diff
@@ -122,16 +122,71 @@ pub struct Ipv6Addr {
     inner: c::in6_addr,
 }
 
-#[allow(missing_docs)]
```
❤️

KodrAus commented on library/std/src/net/ip.rs:
```diff
@@ -186,12 +255,19 @@ impl IpAddr {
+    /// future, as new RFCs are published.
+    ///
+    /// [changes]: ../net/index.html#stability-guarantees-for-ietf-defined-behavior
+
```
```suggestion
#[stable(feature = "ip", since = "1.47.0")]
```

KodrAus commented on library/std/src/net/ip.rs:
```diff
@@ -238,6 +320,14 @@ impl IpAddr {
+    /// future, as new RFCs are published.
+    ///
+    /// [changes]: ../net/index.html#stability-guarantees-for-ietf-defined-behavior
+
```
```suggestion
#[stable(feature = "ip", since = "1.47.0")]
```

KodrAus commented on library/std/src/net/ip.rs:
```diff
@@ -526,6 +657,14 @@ impl Ipv4Addr {
+    /// future, as new RFCs are published.
+    ///
+    /// [changes]: ../net/index.html#stability-guarantees-for-ietf-defined-behavior
+
```
```suggestion
#[stable(feature = "ip", since = "1.47.0")]
```

KodrAus commented on library/std/src/net/ip.rs:
```diff
@@ -563,6 +702,14 @@ impl Ipv4Addr {
+    /// future, as new RFCs are published.
+    ///
+    /// [changes]: ../net/index.html#stability-guarantees-for-ietf-defined-behavior
+
```
```suggestion
#[stable(feature = "ip", since = "1.47.0")]
```

KodrAus commented on library/std/src/net/ip.rs:
```diff
@@ -595,6 +742,14 @@ impl Ipv4Addr {
+    /// future, as new RFCs are published.
+    ///
+    /// [changes]: ../net/index.html#stability-guarantees-for-ietf-defined-behavior
+
```
```suggestion
#[stable(feature = "ip", since = "1.47.0")]
```

KodrAus commented on library/std/src/net/ip.rs:
```diff
@@ -618,6 +773,14 @@ impl Ipv4Addr {
+    /// future, as new RFCs are published.
+    ///
+    /// [changes]: ../net/index.html#stability-guarantees-for-ietf-defined-behavior
+
```
```suggestion
#[stable(feature = "ip", since = "1.47.0")]
```

KodrAus commented on library/std/src/net/ip.rs:
```diff
@@ -650,6 +813,14 @@ impl Ipv4Addr {
+    /// future, as new RFCs are published.
+    ///
+    /// [changes]: ../net/index.html#stability-guarantees-for-ietf-defined-behavior
+
```
```suggestion
#[stable(feature = "ip", since = "1.47.0")]
```

KodrAus commented on library/std/src/net/ip.rs:
```diff
@@ -1169,13 +1396,20 @@ impl Ipv6Addr {
+    /// future, as new RFCs are published.
+    ///
+    /// [changes]: ../net/index.html#stability-guarantees-for-ietf-defined-behavior
+
```
```suggestion
#[stable(feature = "ip", since = "1.47.0")]
```

KodrAus commented on library/std/src/net/ip.rs:
```diff
@@ -1195,12 +1429,19 @@ impl Ipv6Addr {
+    /// future, as new RFCs are published.
+    ///
+    /// [changes]: ../net/index.html#stability-guarantees-for-ietf-defined-behavior
+
```
```suggestion
#[stable(feature = "ip", since = "1.47.0")]
```

KodrAus commented on library/std/src/net/ip.rs:
```diff
@@ -1255,6 +1495,15 @@ impl Ipv6Addr {
+    /// future, as new RFCs are published.
+    ///
+    /// [changes]: ../net/index.html#stability-guarantees-for-ietf-defined-behavior
+
```
```suggestion
#[stable(feature = "ip", since = "1.47.0")]
```

KodrAus commented on library/std/src/net/ip.rs:
```diff
@@ -1311,6 +1559,15 @@ impl Ipv6Addr {
+    /// future, as new RFCs are published.
+    ///
+    /// [changes]: ../net/index.html#stability-guarantees-for-ietf-defined-behavior
+
```
```suggestion
#[stable(feature = "ip", since = "1.47.0")]
```

KodrAus commented on library/std/src/net/ip.rs:
```diff
@@ -1350,6 +1606,14 @@ impl Ipv6Addr {
+    /// future, as new RFCs are published.
+    ///
+    /// [changes]: ../net/index.html#stability-guarantees-for-ietf-defined-behavior
+
```
```suggestion
#[stable(feature = "ip", since = "1.47.0")]
```

KodrAus commented on library/std/src/net/ip.rs:
```diff
@@ -1366,12 +1630,19 @@ impl Ipv6Addr {
+    /// future, as new RFCs are published.
+    ///
+    /// [changes]: ../net/index.html#stability-guarantees-for-ietf-defined-behavior
+
```
```suggestion
#[stable(feature = "ip", since = "1.47.0")]
```

KodrAus commented on library/std/src/net/ip.rs:
```diff
@@ -1401,12 +1672,19 @@ impl Ipv6Addr {
+    /// future, as new RFCs are published.
+    ///
+    /// [changes]: ../net/index.html#stability-guarantees-for-ietf-defined-behavior
+
```
```suggestion
#[stable(feature = "ip", since = "1.47.0")]
```

KodrAus commented on library/std/src/net/ip.rs:
```diff
@@ -1432,19 +1709,32 @@ impl Ipv6Addr {
+    /// future, as new RFCs are published.
+    ///
+    /// [changes]: ../net/index.html#stability-guarantees-for-ietf-defined-behavior
+
```
```suggestion
#[stable(feature = "ip", since = "1.47.0")]
```

KodrAus commented on library/std/src/net/ip.rs:
````diff
@@ -1483,14 +1780,14 @@ impl Ipv6Addr {
     ///            Some(Ipv4Addr::new(192, 10, 2, 255)));
     /// assert_eq!(Ipv6Addr::new(0, 0, 0, 0, 0, 0, 0, 1).to_ipv4_mapped(), None);
     /// ```
+
````
```suggestion
#[stable(feature = "ip", since = "1.47.0")]
```

KodrAus commented on library/std/src/net/ip.rs:
````diff
@@ -186,12 +255,19 @@ impl IpAddr {
     ///
     /// ```
     /// #![feature(ip)]
````
```suggestion

```

KodrAus commented on library/std/src/net/ip.rs:
````diff
@@ -229,7 +312,6 @@ impl IpAddr {
     ///
     /// ```
     /// #![feature(ip)]
````
```suggestion

```

KodrAus commented on library/std/src/net/ip.rs:
````diff
@@ -482,7 +614,6 @@ impl Ipv4Addr {
     ///
     /// ```
     /// #![feature(ip)]
````
```suggestion

```

KodrAus commented on library/std/src/net/ip.rs:
````diff
@@ -1366,12 +1630,19 @@ impl Ipv6Addr {
     ///
     /// ```
     /// #![feature(ip)]
````
```suggestion

```

KodrAus commented on library/std/src/net/ip.rs:
````diff
@@ -1401,12 +1672,19 @@ impl Ipv6Addr {
     ///
     /// ```
     /// #![feature(ip)]
````
```suggestion

```

KodrAus commented on library/std/src/net/ip.rs:
````diff
@@ -1417,13 +1695,12 @@ impl Ipv6Addr {
     /// # Examples
     ///
     /// ```
     /// #![feature(ip)]
````
```suggestion

```

KodrAus commented on library/std/src/net/ip.rs:
````diff
@@ -1483,14 +1780,14 @@ impl Ipv6Addr {
     ///
     /// ```
     /// #![feature(ip)]
````
```suggestion

```

KodrAus commented on library/std/src/net/ip.rs:
````diff
@@ -1169,13 +1396,20 @@ impl Ipv6Addr {
     ///
     /// ```
     /// #![feature(ip)]
````
```suggestion

```

KodrAus commented on library/std/src/net/ip.rs:
````diff
@@ -1195,12 +1429,19 @@ impl Ipv6Addr {
     ///
     /// ```
     /// #![feature(ip)]
````
```suggestion

```

KodrAus commented on library/std/src/net/ip.rs:
````diff
@@ -1227,7 +1468,6 @@ impl Ipv6Addr {
     ///
     /// ```
     /// #![feature(ip)]
````
```suggestion

```

KodrAus commented on library/std/src/net/ip.rs:
````diff
@@ -1285,7 +1534,6 @@ impl Ipv6Addr {
     ///
     /// ```
     /// #![feature(ip)]
````
```suggestion

```

KodrAus commented on library/std/src/net/ip.rs:
````diff
@@ -1333,7 +1590,6 @@ impl Ipv6Addr {
     ///
     /// ```
     /// #![feature(ip)]
````
```suggestion

```

KodrAus reviewed (commented):
Thanks for taking the time to put this together @caass!

I've left some suggestions for adding the right `#[stable]` attributes we need to actually stabilize this API. It's not visible from the diff, but at the top of `ip.rs` there's also this attribute:

```rust
#![unstable(
    feature = "ip",
    reason = "extra functionality has not been \
                                      scrutinized to the level that it should \
                                      be to be stable",
    issue = "27709"
)]
```

We'll need to remove this along with adding the `#[stable]` attributes.

KodrAus commented:
> I'd personally go ahead with just the one is_unicast_link_local that validates against /64, since it's a more strict subset of /10, and if someone runs into trouble with running custom link-local addresses outside of /64 space I kind of feel like that is a problem outside the scope of the Rust programming language, and is something they should talk to their network administrator about.

This sounds like a good path forward to me! Thanks for digging through and making sense of all this material ❤️

caass commented:
Ok, so I went through and added checks for secret hidden IPv4 addresses to all (i think -- i may have missed some) of the IPv6 checks. I also renamed `is_unicast_link_local_strict` to `is_unicast_link_local` and removed the old `is_unicast_link_local`. Is there anything else I should do?

bors commented:
:umbrella: The latest upstream changes (presumably #76422) made this pull request unmergeable. Please resolve the merge conflicts.

KodrAus commented on library/std/src/net/ip.rs:
```diff
@@ -1131,10 +1339,22 @@ impl Ipv6Addr {
     #[stable(since = "1.7.0", feature = "ip_17")]
     pub const fn is_unspecified(&self) -> bool {
-        u128::from_be_bytes(self.octets()) == u128::from_be_bytes(Ipv6Addr::UNSPECIFIED.octets())
+        if let Some(v4_addr) = self.to_ipv4() {
```
I don't _think_ this is a behavioral change, because an unspecified address is also just `0.0.0.0`.

KodrAus commented on library/std/src/net/ip.rs:
```diff
@@ -1151,10 +1371,21 @@ impl Ipv6Addr {
     #[stable(since = "1.7.0", feature = "ip_17")]
     pub const fn is_loopback(&self) -> bool {
-        u128::from_be_bytes(self.octets()) == u128::from_be_bytes(Ipv6Addr::LOCALHOST.octets())
+        if let Some(v4_addr) = self.to_ipv4() {
```
Documenting that this is a behavioral breaking change for:

- `::127.0.0.1` / `::ffff:127.0.0.1`

KodrAus commented on library/std/src/net/ip.rs:
```diff
@@ -1464,10 +1698,21 @@ impl Ipv6Addr {
     #[stable(since = "1.7.0", feature = "ip_17")]
     pub const fn is_multicast(&self) -> bool {
-        (self.segments()[0] & 0xff00) == 0xff00
+        if let Some(v4_addr) = self.to_ipv4() {
```
Documenting that this is a behavioral breaking change for:

- `::224.254.0.0` / `::ffff:224.254.0.0`
- `::236.168.10.65` / `::ffff:236.168.10.65`

KodrAus reviewed (commented):
Thanks @caass!

I think the merge conflicts will be coming from this change: https://github.com/rust-lang/rust/commit/a2e077e405f9538b42d32e18082d50d0102d2334 They've just moved inline to the `ip` module.

It looks like we've got a failing UI test in here:

```
failures:

---- [ui] ui/consts/std/net/ipv6.rs stdout ----

error: test run failed!
status: exit code: 101
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/consts/std/net/ipv6/a"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
thread 'main' panicked at 'assertion failed: IS_LOOPBACK', /checkout/src/test/ui/consts/std/net/ipv6.rs:21:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

------------------------------------------
```

It might be easier to figure out what test this is after a rebase.

JohnCSimon commented:
Ping from triage
@caass - Can you please address the build failures and merge conflicts, thank you.

caass commented:
> @caass - Can you please address the build failures and merge conflicts, thank you.

@JohnCSimon hi! I have been sitting here hoping someone with more knowledge of macros than I will be able to write/fix the test cases before i keep pushing on this PR, but I suppose eventually I'll have to learn them myself 😅 I'll resolve the merge conflicts, but the tests are broken until I (or someone else 🤞) rewrite them to reflect the behavioral changes mentioned by @KodrAus :)

little-dude commented:
@caass I can probably help with that.

caass commented:
> @caass I can probably help with that.

!!! @little-dude i would love that! i'll merge later today and write some comments talking about what exactly needs to be done :)

caass commented:
Hello all! A quick summary of the commits since we last talked:

- [285d1fb](https://github.com/rust-lang/rust/pull/76098/commits/285d1fb8649f531fd520d560e913ce6518e72955) and [
66f1cf1](https://github.com/rust-lang/rust/pull/76098/commits/66f1cf1d4440864514f308b00fb3fe1276ea046e) resolved some merge conflicts due to file restructuring and usage of `const fn` on master.
- [5584599](https://github.com/rust-lang/rust/pull/76098/commits/55845998f95212aea4b4f139db7581284c547768) did, in fact, break tests that were checking for the more relaxed behavior of `is_unicast_link_local` that we decided to remove, as per [our discussion](https://github.com/rust-lang/rust/pull/76098#issuecomment-686901980).
- [f04ea19](https://github.com/rust-lang/rust/pull/76098/commits/f04ea19eb21d168424bb54f5f36fa7efd83c510d) flipped the behavior of checks for hidden IPv4 addresses, so that now IPv6 addresses will follow the following pattern:
```rust
pub const fn is_something(&self) -> bool {
    // IPv6 checking
    self.octets()[0] == IPv6::SOMETHING.octets()[0] ||
    if let Some(ipv4_addr) = self.to_ipv4() { ipv4_addr.is_something() } else { false }
```
This means that now, methods called on IPv6 addresses will correctly check against IPv6 logic before deferring to checking for hidden IPv4 addresses, which fixed a couple tests.
- [e0f66e4](https://github.com/rust-lang/rust/pull/76098/commits/e0f66e43da61d8fe8aadc6bcdc2545de51607332) fixes the remainder of the tests, which were checks against the more relaxed behavior (as mentioned above). Now, they check for the inverse conditions, meaning the tests were reused but with opposite parameters.

Pinging @KodrAus, but anyone else is more than welcome to give feedback if I've messed anything up or missed anything. Particularly with the test cases! ❤️

caass commented:
Hi! Another update -- after some [discussion on zulip](https://rust-lang.zulipchat.com/#narrow/stream/212497-t-compiler.2Fhelp-wanted/topic/fix.20tests.20for.20.2376098/near/212613062) @NeoRaider mentioned that the route we've gone is inconsistent with other similar implementations:

> A later RFC also specifies the whole of fe80::/10 for link-local unicast: https://tools.ietf.org/html/rfc5156#section-2.4
> FWIW, this is also what the Linux kernel implements, and IN6_IS_ADDR_LINKLOCAL() of the C standard library

I feel like it's better to do the thing that more people do rather than the other way around, and revert the behavior of `is_unicast_link_local` to its original looser implementation. I personally would also just *remove* `is_unicast_link_local_strict`, as it has imo more potential to cause confusion than benefit to users.

little-dude commented:
FWIW [go also uses fe80::/10](https://github.com/golang/go/blob/9449a125e8b25d21c0d522435be44a4a6e7af2d3/src/net/ip.go#L157):

```go
// IsLinkLocalUnicast reports whether ip is a link-local
// unicast address.
func (ip IP) IsLinkLocalUnicast() bool {
	if ip4 := ip.To4(); ip4 != nil {
		return ip4[0] == 169 && ip4[1] == 254
	}
	return len(ip) == IPv6len && ip[0] == 0xfe && ip[1]&0xc0 == 0x80
}
```

[Python too](https://github.com/python/cpython/blob/3.9/Lib/ipaddress.py#L2249):

```python
_linklocal_network = IPv6Network('fe80::/10')
```

bors commented:
:umbrella: The latest upstream changes (presumably #78127) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

caass commented:
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author

KodrAus commented:
Thanks for working on this and following through all the twists and turns @caass! I think we need a bit of an IP experts list to help rationalize this, but it sounds like we're close.

I'll nominate for the next Libs meeting.

KodrAus commented:
We discussed this at the Libs meeting and felt that the best way forward would be to survey all of these IP methods and compare their behaviour to their counterparts in other languages. This is mostly an issue for v6 addresses, but for cases where it's not obvious enough from the RFCs themselves what the behaviour should be that should give us a reasonable implementation to reference.

Some implementations for us to compare could be:

- [Rust's `std::net::Ipv6Addr` (ours)](https://doc.rust-lang.org/std/net/struct.Ipv6Addr.html)
- [Go's `net.IP`](https://golang.org/pkg/net/#IP)
- [Java's `java.net.Inet6Address`](https://docs.oracle.com/javase/10/docs/api/java/net/Inet6Address.html)
- [.NET's `System.Net.IPAddress`](https://docs.microsoft.com/en-us/dotnet/api/system.net.ipaddress?view=net-5.0)

Once we're happy with how our implementation aligns with these existing ones I think we'll have a good basis to stabilize!

KodrAus commented:
Started sketching out a document here: https://hackmd.io/VrIxyy8xSqeqQn3RdJhupQ

KodrAus commented:
I'm going to spend some time documenting our methods compared to .NET, and then try Go (which is a language I'm not very familiar with yet). If anybody would like to jump in and help out with other languages that would be really helpful!

caass commented:
Got through `Ipv6MulticastScope`, would love a double check to make sure I didn't miss anything but wouldn't blame anyone for *not* doing that.

bors commented:
:umbrella: The latest upstream changes (presumably #76226) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

CDirkx commented on src/test/ui/consts/std/net/ipv4.rs:
```diff
@@ -0,0 +1,57 @@
+    assert!(IS_LOOPBACK);
+
+    const IS_PRIVATE : bool = IP_ADDRESS.is_private();
+    assert!(!IS_PRIVATE);
+
+    const IS_LINK_LOCAL : bool = IP_ADDRESS.is_link_local();
+    assert!(!IS_LINK_LOCAL);
+
+    const IS_GLOBAL : bool = IP_ADDRESS.is_global();
+    assert!(!IS_GLOBAL);
+
+    const IS_SHARED : bool = IP_ADDRESS.is_shared();
+    assert!(!IS_SHARED);
+
+    const IS_IETF_PROTOCOL_ASSIGNMENT : bool = IP_ADDRESS.is_ietf_protocol_assignment();
+    assert!(!IS_IETF_PROTOCOL_ASSIGNMENT);
+
+    const IS_BENCHMARKING : bool = IP_ADDRESS.is_benchmarking();
+    assert!(!IS_BENCHMARKING);
+
+    const IS_RESERVED : bool = IP_ADDRESS.is_reserved();
+    assert!(!IS_RESERVED);
+
+    const IS_MULTICAST : bool = IP_ADDRESS.is_multicast();
+    assert!(!IS_MULTICAST);
+
+    const IS_BROADCAST : bool = IP_ADDRESS.is_broadcast();
+    assert!(!IS_BROADCAST);
+
+    const IS_DOCUMENTATION : bool = IP_ADDRESS.is_documentation();
+    assert!(!IS_DOCUMENTATION);
+
+    const IP_V6_COMPATIBLE : Ipv6Addr = IP_ADDRESS.to_ipv6_compatible();
+    assert_eq!(IP_V6_COMPATIBLE,
+        Ipv6Addr::from([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 127, 0, 0, 1]));
+
+    const IP_V6_MAPPED : Ipv6Addr = IP_ADDRESS.to_ipv6_mapped();
+    assert_eq!(IP_V6_MAPPED,
+        Ipv6Addr::from([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 127, 0, 0, 1]));
+}
```
This ui test can be removed, has been replaced by `ipv4_const`: 

https://github.com/rust-lang/rust/blob/068320b39e3e4839d832b3aa71fa910ba170673b/library/std/src/net/ip/tests.rs#L814-L870

CDirkx commented on src/test/ui/consts/std/net/ipv6.rs:
```diff
@@ -0,0 +1,49 @@
+
+    const SEGMENTS : [u16; 8] = IP_ADDRESS.segments();
+    assert_eq!(SEGMENTS, [0 ,0 ,0 ,0 ,0 ,0 ,0, 1]);
+
+    const OCTETS : [u8; 16] = IP_ADDRESS.octets();
+    assert_eq!(OCTETS, [0 ,0 ,0 ,0 ,0 ,0 ,0, 0 ,0 ,0 ,0 ,0 ,0 ,0, 0, 1]);
+
+    const IS_UNSPECIFIED : bool = IP_ADDRESS.is_unspecified();
+    assert!(!IS_UNSPECIFIED);
+
+    const IS_LOOPBACK : bool = IP_ADDRESS.is_loopback();
+    assert!(IS_LOOPBACK);
+
+    const IS_GLOBAL : bool = IP_ADDRESS.is_global();
+    assert!(!IS_GLOBAL);
+
+    const IS_UNIQUE_LOCAL : bool = IP_ADDRESS.is_unique_local();
+    assert!(!IS_UNIQUE_LOCAL);
+
+    const IS_UNICAST_LINK_LOCAL : bool = IP_ADDRESS.is_unicast_link_local();
+    assert!(!IS_UNICAST_LINK_LOCAL);
+
+    const IS_UNICAST_SITE_LOCAL : bool = IP_ADDRESS.is_unicast_site_local();
+    assert!(!IS_UNICAST_SITE_LOCAL);
+
+    const IS_DOCUMENTATION : bool = IP_ADDRESS.is_documentation();
+    assert!(!IS_DOCUMENTATION);
+
+    const IS_UNICAST_GLOBAL : bool = IP_ADDRESS.is_unicast_global();
+    assert!(!IS_UNICAST_GLOBAL);
+
+    const MULTICAST_SCOPE : Option<Ipv6MulticastScope> = IP_ADDRESS.multicast_scope();
+    assert_eq!(MULTICAST_SCOPE, None);
+
+    const IS_MULTICAST : bool = IP_ADDRESS.is_multicast();
+    assert!(!IS_MULTICAST);
+
+    const IP_V4 : Option<Ipv4Addr> = IP_ADDRESS.to_ipv4();
+    assert_eq!(IP_V4.unwrap(), Ipv4Addr::new(0, 0, 0, 1));
+}
```
This ui test can be removed, has been replaced by `ipv6_const`: 

https://github.com/rust-lang/rust/blob/068320b39e3e4839d832b3aa71fa910ba170673b/library/std/src/net/ip/tests.rs#L872-L920

CDirkx commented on library/std/src/net/mod.rs:
```diff
@@ -14,6 +14,34 @@
+//! globally routable by an RFC, then in a future release [`Ipv4Addr::is_reserved()`] will return
+//! `false` for such addresses, while [`Ipv4Addr::is_global`] will return true.
+//!
+//! [`IpAddr`]: ../../std/net/enum.IpAddr.html
```
With the new rustdoc functionality of 1.48.0, these references can be written as
```
//! [`IpAddr`]: std::net::IpAddr
```

jyn514 replied:
`crate::net::IpAddr`, crates don't know their own name. Thanks for the catch :)

caass replied:
I gave it a shot -- did I do it right?

jyn514 replied:
Looks great :) if you want to make the links even shorter you can use `IpAddr` instead of `crate::net::IpAddr`, I didn't realize it was in scope. Your current code will work just fine though.

CDirkx commented on library/std/src/net/ip.rs:
```diff
@@ -1153,10 +1369,18 @@ impl Ipv6Addr {
     #[stable(since = "1.7.0", feature = "ip_17")]
     pub const fn is_loopback(&self) -> bool {
         u128::from_be_bytes(self.octets()) == u128::from_be_bytes(Ipv6Addr::LOCALHOST.octets())
+            || if let Some(v4_addr) = self.to_ipv4() { v4_addr.is_loopback() } else { false }
```
Regarding the ipv4 in ipv6 concern by @KodrAus:

Using `to_ipv4` causes Rust to consider both ipv4-mapped and ipv4-compatible addresses, i.e. `::ffff:127.0.0.1` and `::127.0.0.1`. This is inconsistent with other languages:

| Language | ipv4-mapped | ipv4-compatible | demo |
| --- | --- | --- | --- |
| Go | ✔️ | ❌  | https://play.golang.org/p/dqDfRtz0AGW |
| C# | ✔️ (since .NET 5)| ❌  | https://dotnetfiddle.net/hhYXgK |
| Java | ✔️ | ❌  | https://ideone.com/2U0peb |

IPv4-Compatible IPv6 address have also been officially deprecated for a long time https://github.com/rust-lang/rust/pull/75150#issuecomment-668680333, therefore I think it would be better to use `to_ipv4_mapped` everywhere.

caass replied:
Would the best way forward to create a `to_ipv4_compatible` method that is not called anywhere in checking ipv4-in-ipv6 stuff, or should we just disregard IPv4-compatible addresses entirely?

caass commented:
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author

resolved merge conflict 🙃

CDirkx commented on library/std/src/net/ip.rs:
```diff
@@ -1135,10 +1343,18 @@ impl Ipv6Addr {
     /// assert_eq!(Ipv6Addr::new(0, 0, 0, 0, 0, 0xffff, 0xc00a, 0x2ff).is_unspecified(), false);
```
With the change to recognize ipv4 in ipv6, we could add a testcase demonstrating this:

```rust
/// also works with IPv4-mapped addresses:
/// assert_eq!(Ipv6Addr::new(0, 0, 0, 0, 0, 0xffff, 0, 0).is_unspecified(), true)
```
And similar testcases for all the other changed `Ipv6Addr` methods.

jyn514 commented on library/std/src/net/ip.rs:
```diff
@@ -1724,43 +1740,6 @@ impl Ipv6Addr {
-    /// [changes]: ../net/index.html#stability-guarantees-for-ietf-defined-behavior
-    #[rustc_const_unstable(feature = "const_ipv6", issue = "76205")]
-    #[stable(feature = "rust1", since = "1.0.0")]
-    pub const fn to_ipv4(&self) -> Option<Ipv4Addr> {
```
This is a stable function, you can't delete it.

caass replied:
yikes ok

caass replied:
what about `to_ipv4_mapped` is that stable?

caass replied:
because otherwise we're going to have two functions doing the same thing :'(

caass replied:
it appears to not be in stable! https://github.com/rust-lang/rust/blob/master/library/std/src/net/ip.rs#L1495-L1503

jyn514 replied:
Ah hmm, it's weird that's missing a stability attribute, apparently that means it's [unstable by default](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2b485b88c6be8d134135f33e2f5a6694). So yeah, it seems fine to deprecate or remove, you probably want @KodrAus to give an opinion though.

caass replied:
Ok. For the time being, I adjusted the behavior of `to_ipv4` to ignore IPv4-Compatible (the deprecated one) addresses and instead to only look for IPv4-Mapped (the currently-in-use-one) behavior. I kind of feel like I'm just wildly changing behavior so we should probably document this as a breaking change for checking for IPv4-Compatible addresses if this is the behavior we wind up going with.

KodrAus replied:
@jyn514 That would be the `#![unstable]` at the top of the module... Which catches me all the time 🙂

KodrAus replied:
It looks like we’ve made a few behavioral changes, but they’re all around consistency of mapped V4 addresses so I think we can roll that up into the stabilization here and let the rest of Libs weigh in on it.

KodrAus replied:
Another option for `to_ipv4` could be to deprecate it in favor of `to_ipv4_mapped` and stabilize that. That might be preferable here.

jyn514 commented on library/std/src/net/ip.rs:
```diff
@@ -122,16 +114,81 @@ pub struct Ipv6Addr {
+/// This enum [may by subject to changes][changes] in the
+/// future, as new RFCs are published.
+///
+/// [changes]: ../net/index.html#stability-guarantees-for-ietf-defined-behavior
```
```suggestion
/// [changes]: super#stability-guarantees-for-ietf-defined-behavior
```

jyn514 commented on library/std/src/net/ip.rs:
```diff
@@ -148,6 +205,13 @@ impl IpAddr {
+    /// This method's behavior [may by subject to changes][changes] in the
+    /// future, as new RFCs are published.
+    ///
+    /// [changes]: ../net/index.html#stability-guarantees-for-ietf-defined-behavior
```
```suggestion
    /// [changes]: super#stability-guarantees-for-ietf-defined-behavior
```

jyn514 commented on library/std/src/net/ip.rs:
```diff
@@ -122,16 +114,81 @@ pub struct Ipv6Addr {
+/// # Stability guarantees
+///
+/// This enum [may by subject to changes][changes] in the
+/// future, as new RFCs are published.
```
Maybe this should clarify 'IETF RFCs'? At first I thought you meant Rust RFCs and was like well of course :laughing:

caass replied:
lol good point

jyn514 commented on library/std/src/net/ip.rs:
```diff
@@ -170,6 +234,13 @@ impl IpAddr {
+    /// This method's behavior [may by subject to changes][changes] in the
+    /// future, as new RFCs are published.
+    ///
+    /// [changes]: ../net/index.html#stability-guarantees-for-ietf-defined-behavior
```
```suggestion
    /// [changes]: super#stability-guarantees-for-ietf-defined-behavior
```

jyn514 commented on library/std/src/net/ip.rs:
```diff
@@ -187,13 +258,19 @@ impl IpAddr {
+    /// This method's behavior [may by subject to changes][changes] in the
+    /// future, as new RFCs are published.
+    ///
+    /// [changes]: ../net/index.html#stability-guarantees-for-ietf-defined-behavior
```
```suggestion
    /// [changes]: self#stability-guarantees-for-ietf-defined-behavior
```

jyn514 reviewed (commented):
(similar comments for the rest of the hard-coded links)

KodrAus commented:
I've started sketching out a little tool we could use to finish filling out our compatibility table: https://github.com/KodrAus/ipcheck

So far it's just a bit of Rust and .NET but will flesh it out for some other languages too.

caass commented:
> I've started sketching out a little tool we could use to finish filling out our compatibility table: https://github.com/KodrAus/ipcheck
> 
> So far it's just a bit of Rust and .NET but will flesh it out for some other languages too.

@KodrAus This seems waaaaay more fun than what I was doing lol. I'll play around with this later today -- should speed us up tremendously. Great idea!

caass commented:
Something just occurred to me -- does it make sense to have an enum for `Ipv6UnicastScope`? We have utility methods for unicast while there's an enum for multicast:
```rust
enum Ipv6MulticastScope { ... }

impl Ipv6Addr {
    fn is_unicast_link_local(&self) -> bool { ... }

    fn is_unicast_site_local(&self) -> bool { ... }

    fn is_unicast_global(&self) -> bool { ... }

    // other methods...

    fn multicast_scope(&self) -> Ipv6MulticastScope { ... }
}
```

To me it makes sense to change this into an enum before everything gets stabilized...thoughts anyone?

jyn514 commented:
+1 to using types instead of booleans, rustc does the same thing for complicated ideas: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_trait_selection/traits/enum.TraitQueryMode.html

caass commented:
I definitely am running out of steam on this PR, and would love for someone to take the lead on it. Pretty sure my sanity dwindles in O(n^2) to the amount of IETF RFCs I read 😖. If anyone reading this is interested in picking up where I left off, please let me know and I'll gladly answer any questions about what's been done and what needs doing still 😄

crlf0710 commented:
@caass Ping from triage: What's next steps here?

bors commented:
:umbrella: The latest upstream changes (presumably #79342) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

caass commented:
@crlf0710 hi triage!

Next steps are testing to ensure this implementation is consistent not only with IETF RFCs but also other popular languages, and then hopefully that's it :)

CDirkx commented:
With #79342 all other ip methods were stabilized as const, so I think all the methods in this PR can also be made const stable.

@caass regarding the comparison with other languages, I recently contributed a bit of functionality to the tool @KodrAus was working on. When the new year starts I'll have quite some time available, let me know if I can help in any way.

jyn514 commented on library/std/src/net/ip.rs:
```diff
@@ -212,7 +212,7 @@ impl IpAddr {
     ///
     /// [changes]: super#stability-guarantees-for-ietf-defined-behavior
-    #[rustc_const_unstable(feature = "const_ip", issue = "76205")]
+    #[rustc_const_stable(feature = "const_ip", since = "1.50.0")]
```
The 1.49 release is in two days, so this should be 1.51 instead. (nightly already shows a 1.51 version number: https://github.com/rust-lang/rust/pull/80366)

jyn514 replied:
Oh wait sorry, this was from a merge commit. Ignore me.

caass replied:
lol

KodrAus commented:
Thanks for your awesome work on that @CDirkx! And @caass for sticking through this marathon.

I just pushed some updates to the tool to include more methods and output a markdown table for us. Here are the results so far:

| addr | .NET | Python | Go | Java |
| - | - | - | - | - |
| `0.0.0.0` | ❌ `{ to_ipv6 : "::ffff:0.0.0.0" (rust) ≠ "::ffff:0:0" }` | ✔️ | ✔️ | ✔️ |
| `0.1.2.3` | ✔️ | ✔️ | ❌ `{ is_global : false (rust) ≠ true }` | ✔️ |
| `1.1.1.1` | ✔️ | ✔️ | ✔️ | ✔️ |
| `127.0.0.1` | ✔️ | ✔️ | ✔️ | ✔️ |
| `255.255.255.255` | ✔️ | ❌ `{ is_reserved : false (rust) ≠ true }` | ✔️ | ✔️ |
| `10.0.0.1` | ✔️ | ✔️ | ❌ `{ is_global : false (rust) ≠ true }` | ✔️ |
| `16.89.10.65` | ✔️ | ✔️ | ✔️ | ✔️ |
| `45.22.13.197` | ✔️ | ✔️ | ✔️ | ✔️ |
| `100.64.0.0` | ✔️ | ✔️ | ❌ `{ is_global : false (rust) ≠ true }` | ✔️ |
| `100.128.0.0` | ✔️ | ✔️ | ✔️ | ✔️ |
| `169.254.0.0` | ❌ `{ is_unicast_link_local : true (rust) ≠ false }` | ✔️ | ✔️ | ✔️ |
| `172.16.10.10` | ✔️ | ✔️ | ❌ `{ is_global : false (rust) ≠ true }` | ✔️ |
| `192.0.0.8` | ✔️ | ❌ `{ is_global : false (rust) ≠ true }` | ❌ `{ is_global : false (rust) ≠ true }` | ✔️ |
| `192.0.0.9` | ✔️ | ✔️ | ✔️ | ✔️ |
| `192.0.0.255` | ✔️ | ❌ `{ is_global : false (rust) ≠ true }` | ❌ `{ is_global : false (rust) ≠ true }` | ✔️ |
| `192.168.0.2` | ✔️ | ✔️ | ❌ `{ is_global : false (rust) ≠ true }` | ✔️ |
| `198.18.0.0` | ✔️ | ✔️ | ❌ `{ is_global : false (rust) ≠ true }` | ✔️ |
| `203.0.113.6` | ✔️ | ✔️ | ❌ `{ is_global : false (rust) ≠ true }` | ✔️ |
| `240.0.0.0` | ✔️ | ✔️ | ❌ `{ is_global : false (rust) ≠ true }` | ✔️ |
| `255.255.255.254` | ✔️ | ✔️ | ❌ `{ is_global : false (rust) ≠ true }` | ✔️ |
| `::ffff:0.0.0.0` | ❌ `{ to_ipv6 : "::ffff:0.0.0.0" (rust) ≠ "::ffff:0:0" }` | ❌ `{ is_global : true (rust) ≠ false, is_reserved : false (rust) ≠ true, is_unspecified : true (rust) ≠ false }` | ❌ `{ is_global : true (rust) ≠ false }` | ✔️ |
| `::ffff:0.1.2.3` | ✔️ | ❌ `{ is_global : true (rust) ≠ false, is_reserved : false (rust) ≠ true }` | ✔️ | ✔️ |
| `::ffff:1.1.1.1` | ✔️ | ❌ `{ is_global : true (rust) ≠ false, is_reserved : false (rust) ≠ true }` | ✔️ | ✔️ |
| `::ffff:127.0.0.1` | ✔️ | ❌ `{ is_global : true (rust) ≠ false, is_loopback : true (rust) ≠ false, is_reserved : false (rust) ≠ true }` | ❌ `{ is_global : true (rust) ≠ false }` | ✔️ |
| `::ffff:255.255.255.255` | ✔️ | ❌ `{ is_global : true (rust) ≠ false, is_reserved : false (rust) ≠ true }` | ❌ `{ is_global : true (rust) ≠ false }` | ✔️ |
| `::ffff:10.0.0.1` | ✔️ | ❌ `{ is_global : true (rust) ≠ false, is_reserved : false (rust) ≠ true }` | ✔️ | ✔️ |
| `::ffff:16.89.10.65` | ✔️ | ❌ `{ is_global : true (rust) ≠ false, is_reserved : false (rust) ≠ true }` | ✔️ | ✔️ |
| `::ffff:45.22.13.197` | ✔️ | ❌ `{ is_global : true (rust) ≠ false, is_reserved : false (rust) ≠ true }` | ✔️ | ✔️ |
| `::ffff:100.64.0.0` | ✔️ | ❌ `{ is_global : true (rust) ≠ false, is_reserved : false (rust) ≠ true }` | ✔️ | ✔️ |
| `::ffff:100.128.0.0` | ✔️ | ❌ `{ is_global : true (rust) ≠ false, is_reserved : false (rust) ≠ true }` | ✔️ | ✔️ |
| `::ffff:169.254.0.0` | ❌ `{ is_unicast_link_local : true (rust) ≠ false }` | ❌ `{ is_global : true (rust) ≠ false, is_reserved : false (rust) ≠ true, is_unicast_link_local : true (rust) ≠ false }` | ❌ `{ is_global : true (rust) ≠ false }` | ✔️ |
| `::ffff:172.16.10.10` | ✔️ | ❌ `{ is_global : true (rust) ≠ false, is_reserved : false (rust) ≠ true }` | ✔️ | ✔️ |
| `::ffff:192.0.0.8` | ✔️ | ❌ `{ is_global : true (rust) ≠ false, is_reserved : false (rust) ≠ true }` | ✔️ | ✔️ |
| `::ffff:192.0.0.9` | ✔️ | ❌ `{ is_global : true (rust) ≠ false, is_reserved : false (rust) ≠ true }` | ✔️ | ✔️ |
| `::ffff:192.0.0.255` | ✔️ | ❌ `{ is_global : true (rust) ≠ false, is_reserved : false (rust) ≠ true }` | ✔️ | ✔️ |
| `::ffff:192.168.0.2` | ✔️ | ❌ `{ is_global : true (rust) ≠ false, is_reserved : false (rust) ≠ true }` | ✔️ | ✔️ |
| `::ffff:198.18.0.0` | ✔️ | ❌ `{ is_global : true (rust) ≠ false, is_reserved : false (rust) ≠ true }` | ✔️ | ✔️ |
| `::ffff:203.0.113.6` | ✔️ | ❌ `{ is_global : true (rust) ≠ false, is_reserved : false (rust) ≠ true }` | ✔️ | ✔️ |
| `::ffff:240.0.0.0` | ✔️ | ❌ `{ is_global : true (rust) ≠ false, is_reserved : false (rust) ≠ true }` | ✔️ | ✔️ |
| `::ffff:255.255.255.254` | ✔️ | ❌ `{ is_global : true (rust) ≠ false, is_reserved : false (rust) ≠ true }` | ✔️ | ✔️ |
| `fdf8:f53b:82e4::53` | ❌ `{ to_ipv4 : null (rust) ≠ "0.0.0.83" }` | ✔️ | ❌ `{ is_global : false (rust) ≠ true }` | ✔️ |
| `fe80::200:5aee:feaa:20a2` | ❌ `{ to_ipv4 : null (rust) ≠ "254.170.32.162" }` | ✔️ | ✔️ | ✔️ |
| `2001:0002:6c::430` | ❌ `{ to_ipv4 : null (rust) ≠ "0.0.4.48" }` | ❌ `{ is_global : true (rust) ≠ false }` | ✔️ | ✔️ |
| `2001:10:240:ab::a` | ❌ `{ to_ipv4 : null (rust) ≠ "0.0.0.10" }` | ❌ `{ is_global : true (rust) ≠ false }` | ✔️ | ✔️ |
| `2002:cb0a:3cdd:1::1` | ❌ `{ to_ipv4 : null (rust) ≠ "0.0.0.1" }` | ✔️ | ✔️ | ✔️ |
| `2001:db8:8:4::2` | ❌ `{ to_ipv4 : null (rust) ≠ "0.0.0.2" }` | ✔️ | ❌ `{ is_global : false (rust) ≠ true }` | ✔️ |
| `ff01:0:0:0:0:0:0:2` | ❌ `{ to_ipv4 : null (rust) ≠ "0.0.0.2" }` | ❌ `{ is_global : false (rust) ≠ true }` | ✔️ | ✔️ |

Operations with only a Rust implementation:
- `mc_scope_realm_local`
- `is_documentation`
- `is_unique_local`
- `mc_scope_admin_local`
- `is_benchmarking`
- `is_shared`
- `is_ietf_protocol_assignment`
- `mc_scope_unassigned`
- `mc_scope_reserved`

KodrAus commented:
@caas dod we want to introduce a `Ipv6UnicastScope` enum? I think that sounds like a good idea too.

KodrAus commented:
Interestingly, it looks like the Rust standard library has the most extensive set of methods on its IP addresses by far. My previous comment lists some of them that don’t appear to have any direct alternative  in any of .NET, Go, Java, or Python.

KodrAus commented:
# Status Report: 2021-01-15

We've been working on this for some time now, but I think we're getting very close to being able to begin a FCP. There are a few current lingering questions to answer:

- Should we replace the `Ipv6Addr::is_unicast_*` methods with a `Ipv6Addr::unicast_scope` method that returns a `Ipv6UnicastScope` enum (https://github.com/rust-lang/rust/pull/76098#issuecomment-735019872)?
- Should we change the behavior of `Ipv6Addr::to_ipv4` to ignore deprecated IPv4-compatible addresses, or deprecate the whole method in favor of the more correct `Ipv6Addr::to_ipv4_mapped` method (https://github.com/rust-lang/rust/pull/76098#discussion_r530465408)?
- Are we ok with `Ipv6Addr::is_*` methods now properly considering mapped (non-deprecated) IPv4 addresses? I'd personally be comfortable considering the old behavior a bug.
- Are there any behavioral differences between other language implementations that we should investigate? (https://github.com/rust-lang/rust/pull/76098#issuecomment-760651861)

CDirkx commented:
Differences between Rust behaviour and other languages (based on https://github.com/rust-lang/rust/pull/76098#issuecomment-760651861):

## IPv4

### Is `192.0.0.8` globally reachable?
Python considers `192.0.0.8` (the IPv4 dummy address) to be globally reachable, seemingly in violation of their own documentation, which (like Rust) refers to the [IANA IPv4 special purpose address registry](https://www.iana.org/assignments/iana-ipv4-special-registry/iana-ipv4-special-registry.xhtml) that explicitly mentions `192.0.0.8` as not globally reachable. Python's `is_global` method was added in Python 3.4 (which came out in 2014), and `192.0.0.8` was reserved in 2015, so this is probably just a bug in their implementation and not an explicit API descision.

### Is `255.255.255.255` reserved?
Python considers `255.255.255.255` to be reserved, Rust does not. From the current docs on `Ipv4Addr::is_reserved`:
> Returns [`true`] if this address is reserved by IANA for future use. [IETF RFC 1112]
> defines the block of reserved addresses as `240.0.0.0/4`. This range normally includes the
>  broadcast address `255.255.255.255`, but this implementation explicitly excludes it, since
>  it is obviously not reserved for future use.

### Which addresses are globally reachable?
Go's implementation of `IsGlobalUnicast` is a lot less strict than Rust's `Ipv4Addr::is_global`. For example, Rust doesn't consider addresses reserved for documentation or private use to be globally reachable, Go does.

## IPv4-mapped IPv6 addresses

### Is `::ffff:0.0.0.0` unspecified?
Python considers `::` to be the only unspecified IPv6 address and `0.0.0.0` to be the only unspecified IPv4 address. This used to be the same in Rust, but now because we delegate with ipv4-mapped ipv6 addresses the current implementation in this PR considers `::ffff:0.0.0.0` to be unspecified.

### Are IPv4-mapped IPv6 addresses globally reachable
Python considers IPv4-mapped IPv6 addresses (`::ffff:x.x.x.x`) to be not globally reachable, Rust does. The [IANA IPv6 special purpose address registry](https://www.iana.org/assignments/iana-ipv6-special-registry/iana-ipv6-special-registry.xhtml) does technically designate these addresses to be not globally reachable, but the entire point of these addresses is to represent (potentially globally reachable) IPv4 addresses in applications.

## IPv6 addresses

### Reserved IPv6 addresses
 Python's `is_reserved` also works with IPv6 addresses, Rust does not have `IPv6::is_reserved` (also not included in this PR).

### Truncating IPv6 addresses when converting to IPv4
.NET handles converting a IPv4-mapped IPv6 address to a IPv4 address the same as Rust, e.g. `::ffff:192.0.0.8` → `192.0.0.8`.
However, .NET also allows converting _any_ IPv6 address to a IPv4 address by just truncating the address, e.g. `2002:cb0a:3cdd:1::192.0.0.8` → `192.0.0.8`.

### Which addresses are globally reachable?
Rust only considers addresses that have a global uni/multicast scope to be globally reachable (and globally reachable IPv4-mapped addresses with this PR). Go only has `IsGlobalUnicast`, no equivalent for multicast. Python considers any address not explicitly designated by the [IANA](https://www.iana.org/assignments/iana-ipv6-special-registry/iana-ipv6-special-registry.xhtml) as not globally reachable as globally reachable. Note that some of these designated addresses do have a global uni/multicast scope, so Rust currently considers them globally reachable, an example is the IPv6 benchmarking addresses: `2001:2::/48`.

CDirkx commented on library/std/src/net/ip.rs:
```diff
@@ -1140,10 +1348,18 @@ impl Ipv6Addr {
+    /// [changes]: super#stability-guarantees-for-ietf-defined-behavior
     #[rustc_const_stable(feature = "const_ipv6", since = "1.50.0")]
     #[stable(since = "1.7.0", feature = "ip_17")]
     pub const fn is_unspecified(&self) -> bool {
         u128::from_be_bytes(self.octets()) == u128::from_be_bytes(Ipv6Addr::UNSPECIFIED.octets())
+            || if let Some(v4_addr) = self.to_ipv4() { v4_addr.is_unspecified() } else { false }
     }
```
Python considers `::` (`Ipv6Addr::UNSPECIFIED`, all bits zero) to be the only unspecified IPv6 address, and `0.0.0.0` to be the only unspecified IPv4 address. I'm not sure what the best approach for Rust here is, as you could indeed argue that `::ffff:0.0.0.0` represents the unspecified `0.0.0.0` address. On the other hand, unspecified == all bits zero makes sense to me, and I don't think `::ffff:0.0.0.0` is defined as unspecified in any protocol.

caass replied:
Ok, I'm on board with that reasoning -- I'll remove the v4 check from this method

CDirkx commented on library/std/src/net/ip.rs:
```diff
@@ -1177,21 +1401,27 @@ impl Ipv6Addr {
     #[rustc_const_unstable(feature = "const_ipv6", issue = "76205")]
     pub const fn is_global(&self) -> bool {
-        match self.multicast_scope() {
+        (match self.multicast_scope() {
             Some(Ipv6MulticastScope::Global) => true,
             None => self.is_unicast_global(),
             _ => false,
-        }
+        }) || if let Some(v4_addr) = self.to_ipv4() { v4_addr.is_global() } else { false }
     }
```
Two comments regarding the [IANA IPv6 special purpose address registry](https://www.iana.org/assignments/iana-ipv6-special-registry/iana-ipv6-special-registry.xhtml):

Technically all IPv4-mapped addresses are designated as not globally reachable. However, the entire point of those addresses is to represent IPv4 addresses. So while e.g. `::ffff:1.1.1.1` is technically not globally reachable, most? applications will treat it as `1.1.1.1`, which is globally reachable. What should Rust do here? 

Also, this implementation ignores some other addresses in the registry designated as not globally reachable, such as IPv6 benchmarking addresses: `2001:2::/48`.

caass replied:
> Technically all IPv4-mapped addresses are designated as not globally reachable. However, the entire point of those addresses is to represent IPv4 addresses. So while e.g. `::ffff:1.1.1.1` is technically not globally reachable, most? applications will treat it as `1.1.1.1`, which is globally reachable. What should Rust do here?

I really don't know one way or the other which option would be "more correct." My gut feeling is that it makes sense to kick the can down the road, and ignore IPv4-mapped addresses in this context, the reason being that if someone is trying to make their IP globally routable they likely won't want it to be IPv4-mapped.

> Also, this implementation ignores some other addresses in the registry designated as not globally reachable, such as IPv6 benchmarking addresses: 2001:2::/48.

What do you think should be done to address this?

KodrAus replied:
If we specifically don't want to forward this one to mapped V4 addresses we could call that out in the documentation with an example that performs the conversion for anybody that does want it.

Otherwise, is it defensible to make these methods all behave consistently with respect to mapped V4 addresses and always forward to them?

CDirkx commented:
To me it makes sense to add a couple more methods to `Ipv6Addr`, for some of the ranges described in the [IANA IPv6 special purpose address registry](https://www.iana.org/assignments/iana-ipv6-special-registry/iana-ipv6-special-registry.xhtml), analogous to those of `Ipv4Addr`:

- `is_ietf_protocol_assignment`
- `is_benchmarking`

`is_documentation`, `is_unique_local` and `is_unicast_link_local` are already added.

Those seem to me like the most relevant ones, maybe `is_ipv4_mapped`? These methods can then also be used in `Ipv6Addr::is_global` to adhere to the registry, see https://github.com/rust-lang/rust/pull/76098#discussion_r558690082.

Dylan-DPC commented:
@KodrAus any updates?

bors commented:
:umbrella: The latest upstream changes (presumably #81596) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #81761) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

caass commented:
Hey, seeing as how functionality to determine if an IP address is unicast site local hasn't been stabilized yet, I'm actually of a mind to not add it. [RFC 3879](https://tools.ietf.org/html/rfc3879) formally deprecates site local addresses, and it makes the logic for unicast scope more confusing as site-local addresses are now considered global.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-9`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1846297439) [(plain)](https://github.com/rust-lang/rust/commit/3ba374bb74db0e73aa925c42efc1ecc5cb9cae71/checks/1846297439/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Suite("src/test/assembly") not skipped for "bootstrap::test::Assembly" -- not in ["src/tools/tidy"]
Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 29 tests
iiiiiiiiiiiiiiiiiiiiiiiiiiiii

 finished in 0.070 seconds
Suite("src/test/incremental") not skipped for "bootstrap::test::Incremental" -- not in ["src/tools/tidy"]
Check compiletest suite=incremental mode=incremental (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
---
Suite("src/test/debuginfo") not skipped for "bootstrap::test::Debuginfo" -- not in ["src/tools/tidy"]
Check compiletest suite=debuginfo mode=debuginfo (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 116 tests
iiiiiiiiii.i.i..i....i...ii.i.i....ii............iiii.......i.....i...i.......ii.i.ii.....iiii.....i 100/116
test result: ok. 78 passed; 0 failed; 38 ignored; 0 measured; 0 filtered out; finished in 2.30s

 finished in 2.378 seconds
Suite("src/test/ui-fulldeps") not skipped for "bootstrap::test::UiFullDeps" -- not in ["src/tools/tidy"]
---
thread '<unnamed>' panicked at 'Box<Any>', library/std/src/thread/tests.rs:205:33
................................................
failures:

---- net::ip::tests::ipv4_mapped stdout ----
thread 'net::ip::tests::ipv4_mapped' panicked at 'assertion failed: mapped_unspecified.is_unspecified()', library/std/src/net/ip/tests.rs:714:5

failures:
    net::ip::tests::ipv4_mapped


test result: FAILED. 847 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 10.33s

error: test failed, to rerun pass '-p std --lib'


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/library/test/Cargo.toml" "-p" "std" "--" "--quiet"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:21:14
```

</details>

caass commented:
> To me it makes sense to add a couple more methods to `Ipv6Addr`, for some of the ranges described in the [IANA IPv6 special purpose address registry](https://www.iana.org/assignments/iana-ipv6-special-registry/iana-ipv6-special-registry.xhtml), analogous to those of `Ipv4Addr`:
> 
>     * `is_ietf_protocol_assignment`
> 
>     * `is_benchmarking`
> 
> 
> `is_documentation`, `is_unique_local` and `is_unicast_link_local` are already added.
> 
> Those seem to me like the most relevant ones, maybe `is_ipv4_mapped`? These methods can then also be used in `Ipv6Addr::is_global` to adhere to the registry, see [#76098 (comment)](https://github.com/rust-lang/rust/pull/76098#discussion_r558690082).

I agree, I'll add those methods -- we've actually removed a `to_ipv4_mapped` method and simply use the `to_ipv4` method instead, which serves the same purpose but is less confusing -- people don't have to learn the difference between ipv4-mapped and ipv4-compatible addresses.

KodrAus commented:
@Dylan-DPC I think our current status here is that we've [identified some differences](https://github.com/rust-lang/rust/pull/76098#issuecomment-761234042) between the behavior of Rust's IP functionality and other languages (some of the .NET ones were me holding it wrong).

We've identified [a few missing methods](https://github.com/rust-lang/rust/pull/76098#issuecomment-761271217) we should add, and are [looking at others](https://github.com/rust-lang/rust/pull/76098#discussion_r558690082) in more detail.

Once we've been through this round of updates I'll collect the complete set of methods and changes together in an update that we can FCP against.

jyn514 commented:
> We've identified a few missing methods we should add, and are looking at others in more detail.

Do those need to be added before stabilizing the feature? It would be backwards compatible to add them later.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1998679023) [(plain)](https://github.com/rust-lang/rust/commit/a1d4e70aa27367f8700370478efb1f35c2d10ae9/checks/1998679023/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
Checking which error codes lack tests...
Found 436 error codes
Found 0 error codes with no tests
Done!
tidy error: `/checkout/src/libstd/net/ip.rs:2098` contains `#[test]`; unit tests and benchmarks must be placed into separate files or directories named `tests.rs`, `benches.rs`, `tests` or `benches`
some tidy checks failed

command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build"
expected success, got: exit code: 1

```

</details>

JohnCSimon commented:
Ping from triage: 
@caass can you please fix the broken build job - thank you!

@rustbot label: -S-waiting-on-review +S-waiting-on-author

Dylan-DPC commented:
r? @dtolnay 

@caass any updates on the failing tests? thanks

lygstate commented:
Would the platform-independent functional to be moved to core?

crlf0710 commented:
@caass Ping from triage, CI is still red. Would you mind fixing the tests? Thanks.

JohnCSimon commented:
Ping from triage: 
@caass - no activity since February 25th, I'm closing this as inactive. Feel free to reopen when you have the chance to continue. Thanks!
